### PR TITLE
GH-37938: [Swift] Add initial C data interface implementation

### PR DESCRIPTION
**Pull request by abandy**
Continuation for PR: #39091

This add an initial implementation of the C Data interface for swift. During development it was found that null count was not being properly maintained on the arrow buffers and this change is included as well. Also some minor refactoring was done to existing sources to enable this feature.

This has been tested from Swift calling into C to import data but not from Swift to C exporting data. Test is currently ongoing.
* GitHub Issue: #37938

### CODE REVIEW

**Comment by abandy** [comment]
The go test for the C data interface are currently not included in the CI tests.  I will update this in a future PR.

**Comment by dbtsai** [comment]
It’s great to add those interfaces to add Swift support. 

Cc @andygrove @viirya 

Thanks!

File: ci/scripts/swift_test.sh [hunk 1]
```diff
@@ -40,4 +40,4 @@ popd
 source_dir=${1}/swift/ArrowFlight
 pushd ${source_dir}
 swift test
-popd
+popd
```
Comment (dbtsai) [inline ci/scripts/swift_test.sh:43]: Unnecessary change

**Comment by abandy** [comment]
> It’s great to add those interfaces to add Swift support.
> 
> Cc @andygrove @viirya
> 
> Thanks!

Np, thank you!

**Comment by kou** [comment]
@dbtsai Are you familiar with Swift? If so, could you also review this carefully?

File: swift/Arrow/Sources/Arrow/ArrowCExporter.swift [hunk 2]
```diff
@@ -0,0 +1,134 @@
...
+
+    public func exportArray(_ cArray: inout ArrowC.ArrowArray, arrowData: ArrowData) {
+        let exportArray = ExportArray(arrowData)
+        cArray.buffers = exportArray.buffers
+        cArray.length = Int64(arrowData.length)
+        cArray.null_count = Int64(arrowData.nullCount)
+        cArray.n_buffers = Int64(arrowData.buffers.count)
+        cArray.n_children = 0
+        cArray.children = nil
+        cArray.dictionary = nil
```
Comment (viirya) [inline swift/Arrow/Sources/Arrow/ArrowCExporter.swift:116]: Hmm, is this for TODO later? Otherwise I think `n_children` should not be zero always? And also for `children` and `dictionary`.

File: swift/Arrow/Sources/Arrow/ArrowCExporter.swift [hunk 3]
```diff
@@ -0,0 +1,134 @@
...
+            let exportSchema = try ExportSchema(arrowType, name: name)
+            cSchema.format = exportSchema.arrowTypeName
+            cSchema.name = exportSchema.name
+            cSchema.private_data =
+                UnsafeMutableRawPointer(mutating: UnsafeRawPointer(bitPattern: exportSchema.id))
+            cSchema.release = {(data: UnsafeMutablePointer<ArrowC.ArrowSchema>?) in
+                let arraySchema = data!.pointee
+                let exportId = Int(bitPattern: arraySchema.private_data)
+                guard ArrowCExporter.exportedData[exportId] != nil else {
+                    fatalError("Export schema not found with id \(exportId)")
```
Comment (viirya) [inline swift/Arrow/Sources/Arrow/ArrowCExporter.swift:88]: Does `fatalError` actually do something like C `exit`?

File: swift/Arrow/Package.swift [hunk 4]
```diff
@@ -36,18 +36,27 @@ let package = Package(
...
+              url: "https://github.com/apple/swift-atomics.git",
+              .upToNextMajor(from: "1.2.0") // or `.upToNextMinor
+            )
     ],
     targets: [
         // Targets are the basic building blocks of a package. A target can define a module or a test suite.
         // Targets can depend on other targets in this package, and on products in packages this package depends on.
+        .target(
+            name: "ArrowC", // your C/C++ library's name
+            path: "Sources/ArrowC" // your path to the C/C++ library
```
[the remark this answers is not in the document]
Comment (abandy) [reply swift/Arrow/Package.swift:50]: Good catch!  I will remove.

File: swift/Arrow/Sources/Arrow/ArrowCExporter.swift (hunk 2 above)
Comment (abandy) [reply swift/Arrow/Sources/Arrow/ArrowCExporter.swift:116]: The Swift version of Arrow currently doesn't support having children (Struct type) or dictionaries.  I will add a comment mentioning this so it can be updated when those constructs are added.

File: swift/Arrow/Sources/Arrow/ArrowCExporter.swift (hunk 3 above)
Comment (abandy) [reply swift/Arrow/Sources/Arrow/ArrowCExporter.swift:88]: I think the behavior is the same.  Description from Swift docs for fatalError: Unconditionally prints a given message and stops execution.

**Comment by abandy** [comment]
@kou and @pitrou Please review when you get a chance.  This is very similar to the previous PR(https://github.com/apache/arrow/pull/39091) for this enhancement that was closed but with the updates from @pitrou previous comments.  Thank you!

File: swift/Arrow/Package.swift
Comment (kou) [inline swift/Arrow/Package.swift:1]: Could you add this file to `swiftlint` targets?

Comment (abandy) [reply swift/Arrow/Package.swift:1]: I also added ArrowFlight and CDataWGo Package files as well, thanks!

**Comment by abandy** [comment]
@kou I have made the requested changes. Please review again.

**Comment by pitrou** [comment]
@abandy Could you please rebase from git main, if not recently done, so that we get cleaner CI results?

File: swift/Arrow/Sources/Arrow/ArrowCExporter.swift [hunk 5]
```diff
@@ -0,0 +1,133 @@
...
+            for arrowBuffer in arrowData.buffers {
+                data.append(arrowBuffer.rawPointer)
+            }
+
+            self.buffers = UnsafeMutablePointer(mutating: data)
+            super.init()
+        }
+    }
+
+    private static var exportedData = [Int: ExportData]()
```
Comment (pitrou) [inline swift/Arrow/Sources/Arrow/ArrowCExporter.swift:70]: Is this container cleaned up at shutdown? I'm wondering what happens if a release callback is called after this container is cleaned up. Presumably the release callback would crash because of the `fatalError`?

This is not necessarily a problem, but I'm curious :-)

File: swift/Arrow/Sources/Arrow/ArrowCExporter.swift [hunk 6]
```diff
@@ -0,0 +1,133 @@
...
+    private class ExportSchema: ExportData {
+        public let arrowTypeName: UnsafePointer<CChar>
+        public let nameCstr: UnsafePointer<CChar>
+        private let arrowType: ArrowType
+        private let name: String
+        init(_ arrowType: ArrowType, name: String = "") throws {
+            self.arrowType = arrowType
+            // keeping the name str to ensure the cstring buffer remains valid
+            self.name = name
+            self.arrowTypeName = (try arrowType.cDataFormatId as NSString).utf8String!
```
Comment (pitrou) [inline swift/Arrow/Sources/Arrow/ArrowCExporter.swift:46]: This is only ok because `arrowType.cDataFormatId` always returns a static string, right?
Otherwise you would also have to store `arrowType.cDataFormatId` in `ExportSchema` so that its buffer remains valid.

File: swift/Arrow/Sources/Arrow/ArrowCImporter.swift [hunk 7]
```diff
@@ -0,0 +1,176 @@
...
+            return .failure(.invalid("Error occurred while attempting to import type: \(error)"))
+        }
+    }
+
+    public func importField(_ cSchema: ArrowC.ArrowSchema) ->
+        Result<ArrowField, ArrowError> {
+        if cSchema.n_children > 0 {
+            return .failure(.invalid("Children currently not supported"))
+        } else if cSchema.dictionary != nil {
+            return .failure(.invalid("Dictinoary types currently not supported"))
```
Comment (pitrou) [inline swift/Arrow/Sources/Arrow/ArrowCImporter.swift:76]: Should also call `ArrowCImporter.release(cSchema)` here?

File: swift/Arrow/Sources/Arrow/ArrowCImporter.swift [hunk 8]
```diff
@@ -0,0 +1,176 @@
...
+                appendToBuffer(cArray.buffers[0], arrowBuffers: &arrowBuffers, length: UInt(ceil(Double(length) / 8)))
+                appendToBuffer(cArray.buffers[1], arrowBuffers: &arrowBuffers, length: length)
+            }
+        }
+
+        switch makeArrayHolder(arrowField, buffers: arrowBuffers, nullCount: nullCount) {
+        case .success(let holder):
+            return .success(ImportArrayHolder(holder, cArrayPtr: cArrayPtr))
+        case .failure(let err):
+            return .failure(err)
```
Comment (pitrou) [inline swift/Arrow/Sources/Arrow/ArrowCImporter.swift:158]: Should also call `ArrowCImporter.release` here?

File: swift/Arrow/Sources/Arrow/ArrowCImporter.swift (hunk 7 above)
Comment (abandy) [reply swift/Arrow/Sources/Arrow/ArrowCImporter.swift:76]: I see, agreed, I will update.

File: swift/Arrow/Sources/Arrow/ArrowCExporter.swift (hunk 6 above)
Comment (abandy) [reply swift/Arrow/Sources/Arrow/ArrowCExporter.swift:46]: Yes, that is correct.

File: swift/Arrow/Sources/Arrow/ArrowCExporter.swift (hunk 5 above)
Comment (abandy) [reply swift/Arrow/Sources/Arrow/ArrowCExporter.swift:70]: Yes, the memory should be released during shutdown.  The container should be valid in the process that loaded the library until the process has been shutdown.

**Comment by abandy** [comment]
@pitrou and @kou I have made the requested changes. Please review again.  I have 2 PRs waiting for some of theses changes (one is for adding struct type) and it would be great if we could get this merged so I can submit them.  Thank you!

File: swift/.swiftlint.yml [hunk 9]
```diff
@@ -16,10 +16,15 @@
 # under the License.
 
 included:
+
```
Comment (kou) [inline swift/.swiftlint.yml:19]:
```suggestion
```

**Review by kou — COMMENTED** [review]
In general, smaller PRs are easy to review/merge than one large PR.
If you want to proceed quickly as much as possible, could you use the smaller PRs approach next time like you will do for struct type?
For example, we might be able to choose the following approach for this case:
1. Implement `ArrowSchema` export for 1 simple type such as `bool`.
2. Implement `ArrowSchema` import for 1.
3. Implement `ArrowSchema` export/import for one more simple type such as `int8`.
4. ...

FYI: MATLAB uses the small PRs approach: https://github.com/apache/arrow/issues?q=is%3Aissue+matlab+%22C+data+interface%22

**Comment by abandy** [comment]
> In general, smaller PRs are easy to review/merge than one large PR. If you want to proceed quickly as much as possible, could you use the smaller PRs approach next time like you will do for struct type? For example, we might be able to choose the following approach for this case:
> 
> 1. Implement `ArrowSchema` export for 1 simple type such as `bool`.
> 2. Implement `ArrowSchema` import for 1.
> 3. Implement `ArrowSchema` export/import for one more simple type such as `int8`.
> 4. ...
> 
> FYI: MATLAB uses the small PRs approach: https://github.com/apache/arrow/issues?q=is%3Aissue+matlab+%22C+data+interface%22

Hi @kou, thank you example on the smaller project.  I will definitely follow that format for the struct type change!  Thank you very much!

File: swift/Arrow/Sources/Arrow/ArrowCExporter.swift (hunk 6 above)
Comment (pitrou) [reply swift/Arrow/Sources/Arrow/ArrowCExporter.swift:46]: Ok, I suppose that's not very future-proof, then? Some types such as timestamp, union... have dynamic format strings which embody the type's parameters. You may want to add a comment to remind readers about this?

(or perhaps simply arrange to keep the `arrowType.cDataFormatId` result alive in the `ExportSchema` class)

File: swift/Arrow/Sources/Arrow/ArrowCExporter.swift (hunk 5 above)
Comment (pitrou) [reply swift/Arrow/Sources/Arrow/ArrowCExporter.swift:70]: Yes, my question was more along the lines of: what happens if the release callback is called _after_ finalization of Swift internals has started? Granted, we might not be able to do much about it, but it's probably good to keep this in mind (and/or add a comment about it).

**Review by pitrou — APPROVED** [review]
@abandy This LGTM in general, I've just posted two comments about potential corner cases. I'll let @kou give the final go.

File: swift/Arrow/Sources/Arrow/ArrowCExporter.swift (hunk 6 above)
Comment (abandy) [reply swift/Arrow/Sources/Arrow/ArrowCExporter.swift:46]: I have a PR ready for Struct types and plan to follow up with Unions afterward.  I was going to address this issue when I added support for these types to the C data interface in a future PR but I will make this change now so it is one less thing that will be an issue at that time.  Thank you!

File: swift/Arrow/Sources/ArrowC/include/ArrowCData.h [hunk 10]
```diff
@@ -0,0 +1,81 @@
...
+// under the License.
+
+#ifndef ARROW_C_DATA_INTERFACE
+#define ARROW_C_DATA_INTERFACE
+
+#define ARROW_FLAG_DICTIONARY_ORDERED 1
+#define ARROW_FLAG_NULLABLE 2
+#define ARROW_FLAG_MAP_KEYS_SORTED 4
+
+#include <stdio.h>
```
Comment (kou) [inline swift/Arrow/Sources/ArrowC/include/ArrowCData.h:25]: Can we remove this?

File: swift/Arrow/Sources/ArrowC/include/ArrowCData.h [hunk 11]
```diff
@@ -0,0 +1,81 @@
...
+#include <stdlib.h>
```
Comment (kou) [inline swift/Arrow/Sources/ArrowC/include/ArrowCData.h:26]: Can we move this to `ArrowCData.c`?

File: swift/Arrow/Sources/ArrowC/include/ArrowCData.h [hunk 12]
```diff
@@ -0,0 +1,81 @@
...
+#include <stdint.h>  // Must have this!
```
Comment (kou) [inline swift/Arrow/Sources/ArrowC/include/ArrowCData.h:27]:
```suggestion
#include <stdint.h>  // For int64_t
```

File: swift/CDataWGo/go.mod [hunk 13]
```diff
@@ -0,0 +1,41 @@
...
+// limitations under the License.
+
+module go-swift
+
+go 1.21
+
+require (
+	github.com/andybalholm/brotli v1.0.5 // indirect
+	github.com/apache/arrow/go/v12 v12.0.1 // indirect
+	github.com/apache/arrow/go/v16 v16.0.0-20240203105949-22f2cfd1e1eb // indirect
```
Comment (kou) [inline swift/CDataWGo/go.mod:24]: Is this really indirect dependency?

**Review by kou — APPROVED** [review]
+1

I added some minor comments but I'll merge this. If we need to work on these comments, let's work on them as follow-up tasks.

**Comment by abandy** [comment]
> +1
> 
> I added some minor comments but I'll merge this. If we need to work on these comments, let's work on them as follow-up tasks.

Thank you @kou, @pitrou, @dbtsai, @viirya for the reviews!! 

I will create a MINOR pr for the comments added by @kou.
